File: Counting_Sort/Ruby/mitogh/counting_sort.rb
def counting_sort(array = [])
  # Search the maximum value
  max = array.max
  max ||= 0

  # Create an array with 0 to max elements with zero
  aux = Array.new(max, 0)

  # Do the count
  array.each do |element|
    aux[element] = aux[element].to_i + 1
  end

  # The new order array
  order = []
  # Fill the new array
  aux.each.with_index do |element, index|
      element.times do 
          order << index
      end
  end
  return order
end

File: Counting_Sort/PHP/mitogh/counting_sort.php
<?php
/**
 * Counting Sort Algirithm 
 *
 * @param   Array   $array  The original array
 * @return  Array   $order  The order array
 */
function counting_sort($array = []){
    // Return [] if the array has no elements
    if( !is_array($array) || !count($array) ){
        return $array;
    }
    
    // Find the min and max value
    $max = max($array);
    $min = min($array);
    
    // Fill the aux array from min to max with 0
    $aux = [];
    $aux = array_fill($min, $max, 0);

    foreach($array as $element){
        $aux[$element]++;
    }

    // Fill the order array
    $order = [];
    foreach($aux as $key => $value){
        $i = 0; 
        while($i++ < $value){
            array_push($order, $key);
        }
    }
    return $order;
}
